browser/hooks/useEvmToSolanaSwap.ts

```diff
@@ -0,0 +1,239 @@
+      const { publicKey, sendTransaction } = solanaWallet;
+      const evmProvider = getEvmProvider(sourceChain);
+
+      /**
+       * STEP 3: Connect to smart contracts
+       */
+      const evmTokenDetails = getTokenDetails(
+        evmChainConfig,
+        sourceTokenProjectId,
+      );
+      const evmTokenContract = ERC20Token__factory.connect(
+        evmTokenDetails.address,
+        evmProvider,
+      );
+      const evmRoutingContract = Routing__factory.connect(
+        EVM_CHAIN_CONFIGS[sourceChain].routingContractAddress,
+        evmSigner,
+      );
+
+      /**
+       * STEP 4: Create SPL token accounts if required
+       * */
+      const userTokenAccounts = await getOrCreateSolanaTokenAccounts(
+        solanaConnection,
+        sendTransaction,
+        publicKey,
+      );
+      logSolanaAccounts("User SPL token accounts", userTokenAccounts);
+
+      /**
+       * STEP 5: Approve ERC20 token spend if required
+       */
+      const inputAmountAtomic = utils.parseUnits(
+        inputAmount,
+        evmTokenDetails.decimals,
+      );
+      const currentApprovalAmountAtomic = await evmTokenContract.allowance(
+        evmAddress,
+        evmChainConfig.routingContractAddress,
+      );
+      if (currentApprovalAmountAtomic.lt(inputAmountAtomic)) {
+        const approvalResponse = await evmTokenContract.approve(
+          evmChainConfig.routingContractAddress,
+          inputAmountAtomic,
+        );
+        console.info(
+          `Source chain approval transaction hash: ${approvalResponse.hash}`,
+        );
+        await approvalResponse.wait();
+      }
+
+      /**
+       * STEP 6: Gather arguments for propeller transfer
+       */
+      const solanaTokenDetails = getTokenDetails(
+        SOLANA_CHAIN_CONFIG,
+        targetTokenProjectId,
+      );
+      const solanaTokenAccount = await getAssociatedTokenAddress(
+        new PublicKey(solanaTokenDetails.address),
+        publicKey,
+      );
+      const solanaOwner = solanaTokenAccount.toBytes();
+      const maxPropellerFeeAtomic = utils.parseUnits(
+        maxPropellerFee,
+        SOLANA_CHAIN_CONFIG.swimUsdDetails.decimals,
+      );
+      const memo = generateId();
+
+      /**
+       * STEP 7: Subscribe to events on source and target chains
+       */
+      const evmFilter = evmRoutingContract.filters.MemoInteraction(
+        bufferToBytesFilter(memo),
+      );
+
+      evmRoutingContract.once(
+        evmFilter,
+        handleEvent("source", sourceChain, (txRecord) => {
+          onTransactionDetected(txRecord);
+        }),
+      );
+
+      const promiseToReturn = new Promise<void>((resolve) => {
+        solanaConnection.onLogs(
+          new PublicKey(SOLANA_CHAIN_CONFIG.routingContractAddress),
+          (logs, context) => {
+            const didFindMemo = logs.logs.find(
+              (log) => log.indexOf(memo.toString("hex")) !== -1,
+            );
+            if (didFindMemo) {
```

Comment:
This will have to be updated once the propeller engine is all working. I'll ping you when I change the script.